### PR TITLE
add device code flow and OBO for MSAL

### PR DESCRIPTION
**Pull request by adamchengtkc**
Device code flow allows auth without browser install on the client. e.g. codespace

`OsduMsalInteractiveCredential(...,devicecodeflow=True)`

MSAL device code flow
https://github.com/Azure-Samples/ms-identity-python-devicecodeflow

### CODE REVIEW

**Review by john-haugland — COMMENTED** [review]
Can you check  if "user_code" in flow before using it ?

**Review by tahir-so — COMMENTED** [review]
I suggest adding a new class.

**Comment by adamchengtkc** [comment]
OsduMsalDeviceCode(...)
No breaking change

**Comment by adamchengtkc** [comment]
Added on-behalf-of flow and bumped version

File: src/osdu/identity/_credential/msal_interactive.py [hunk 1]
```diff
@@ -169,3 +176,76 @@ def refresh_access_token(self) -> dict:
...
+            # the username (or email address) field of the sign-in page for the user,
+            # Often, apps use this parameter during reauthentication,
+            # after already extracting the username from an earlier sign-in
+            # by using the preferred_username claim from returned id_token_claims.
+            # Or simply "select_account" as below - Optional. It forces to show account selector page
+            prompt=msal.Prompt.SELECT_ACCOUNT,
+        )
+
+
+class OsduMsalDeviceCode(OsduMsalInteractiveCredentialBase):
```
Comment (mahewitt) [inline src/osdu/identity/_credential/msal_interactive.py:199]: For consistency should be OsduMsalDeviceCodeCredential

File: src/osdu/identity/_credential/msal_interactive.py [hunk 2]
```diff
@@ -169,3 +176,76 @@ def refresh_access_token(self) -> dict:
...
+        print(self._scopes)
+        flow = app.initiate_device_flow(scopes=[self._scopes])
+        if "user_code" not in flow:
+            print("Failed to get user code for device code flow")
+            print(flow)
+        print(f'Visit https://microsoft.com/devicelogin and enter {flow["user_code"]}')
+        return app.acquire_token_by_device_flow(flow)
+
+
+class OsduMsalOnBehalfOf(OsduBaseCredential):
```
Comment (mahewitt) [inline src/osdu/identity/_credential/msal_interactive.py:213]: For consistency should be OsduMsalOnBehalfOfCredential

**Review by mahewitt — CHANGES_REQUESTED** [review]
Looks good bar the requested naming change for consistency with other Credentials

**Comment by ErlendHaa** [comment]
Hi! Any news on this? Device flow would be really useful :)

**Pull request closed without merging**